File: cpp/s0235.cc
#include <iostream>
#include <vector>

using namespace std;

template <typename T> void print_vec(vector<T> &v) {
  cout << '[' << ' ';
  for (auto i : v)
    cout << i << ' ';
  cout << ']' << '\n';
}

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
  TreeNode *lowestCommonAncestor(TreeNode *root, TreeNode *p, TreeNode *q) {
    if (p->val < root->val && q->val < root->val)
      return lowestCommonAncestor(root->left, p, q);
    else if (p->val > root->val && q->val > root->val)
      return lowestCommonAncestor(root->right, p, q);
    else
      return root;
  }
  TreeNode *lowestCommonAncestor_iter(TreeNode *root, TreeNode *p, TreeNode *q) {
    while (true) {
      if (p->val < root->val && q->val < root->val)
        root = root->left;
      else if (p->val > root->val && q->val > root->val)
        root = root->right;
      else
        return root;
    }
  }
};

int main() {
  std::vector<int> v = {};

  Solution s;

  return 0;
}

File: rust/src/s0235.rs
// Definition for a binary tree node.
pub struct Solution {}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn lowest_common_ancestor(
        root: Option<Rc<RefCell<TreeNode>>>,
        p: Option<Rc<RefCell<TreeNode>>>,
        q: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        let p_val = p.clone().unwrap().borrow().val;
        let q_val = q.clone().unwrap().borrow().val;
        if let Some(node) = root.clone() {
            let mut node = node.borrow_mut();
            if p_val < node.val && q_val < node.val {
                return Self::lowest_common_ancestor(node.left.take(), p, q);
            } else if p_val > node.val && q_val > node.val {
                return Self::lowest_common_ancestor(node.right.take(), p, q);
            } else {
                return root;
            }
        } else {
            None
        }
    }
}
